### PR TITLE
[web] Fixes canvas pixelation and overallocation due to transforms. 

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

- This PR introduces density to PersistedPicture. After Scene preroll, the picture uses the absolute transform to compute best scalar to use to match physical pixels to canvas pixels and rescales recording canvas operations. 
- The scalar (density) is clamped to make sure we don't exceed webkit allocation limits.
- Added code to free up canvas pool when canvas.getContext('2d') fails due to insufficient memory.

## Related Issues

https://github.com/flutter/flutter/issues/59775

## Tests

Added SceneBuilder test that verifies when using a 0.5,0.5 scale transform, a smaller canvas is allocated instead of bounds.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/surface.dart [hunk 1]
```diff
@@ -815,6 +819,13 @@ abstract class PersistedContainerSurface extends PersistedSurface {
 
       // Move the HTML node if necessary.
       if (newChild.rootElement!.parent != childContainer) {
+        if (childContainer == null) {
+          html.Element? elm = childContainer;
+          print(elm);
+        }
+        if (newChild.rootElement == null) {
+          print(newChild);
+        }
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/surface.dart:828]: Leftovers?

File: lib/web_ui/test/engine/surface/scene_builder_test.dart [hunk 2]
```diff
@@ -27,7 +27,7 @@ void testMain() {
     await webOnlyInitializeEngine();
   });
 
-  group('SceneBuilder', () {
+ group('SceneBuilder', () {
```
Comment (yjbanov) [inline lib/web_ui/test/engine/surface/scene_builder_test.dart:30]: undo?

File: lib/web_ui/test/engine/surface/scene_builder_test.dart [hunk 3]
```diff
@@ -525,6 +525,38 @@ void testMain() {
     await testCase('be', 'remove in the middle', deletions: 2);
     await testCase('', 'remove all', deletions: 2);
   });
+
+  test('Canvas should allocate fewer pixels if scaled', () async {
+    final SurfaceSceneBuilder builder = SurfaceSceneBuilder();
+    final Picture picture1 = _drawPicture();
+    builder.pushClipRect(const Rect.fromLTRB(10, 10, 300, 300),);
```
Comment (yjbanov) [inline lib/web_ui/test/engine/surface/scene_builder_test.dart:532]: unnecessary trailing comma

File: lib/web_ui/test/engine/surface/scene_builder_test.dart [hunk 4]
```diff
@@ -525,6 +525,38 @@ void testMain() {
     await testCase('be', 'remove in the middle', deletions: 2);
     await testCase('', 'remove all', deletions: 2);
   });
+
+  test('Canvas should allocate fewer pixels if scaled', () async {
```
Comment (yjbanov) [inline lib/web_ui/test/engine/surface/scene_builder_test.dart:529]: scaled as in "zoomed in" or "zoomed out"? (I know the answer now, but will need to think longer in a few months)

File: lib/web_ui/test/engine/surface/scene_builder_test.dart [hunk 5]
```diff
@@ -525,6 +525,38 @@ void testMain() {
...
+
+    html.HtmlElement content = builder.build().webOnlyRootElement;
+    html.CanvasElement canvas = content.querySelector('canvas');
+    final int unscaledWidth = canvas.width;
+    final int unscaledHeight = canvas.height;
+
+    // Force update to scene which will utilize reuse code path.
+    final SurfaceSceneBuilder builder2 = SurfaceSceneBuilder();
+    builder2.pushOffset(0, 0);
+    builder2.pushTransform(Matrix4.identity().scaled(0.5, 0.5).toFloat64());
```
Comment (yjbanov) [inline lib/web_ui/test/engine/surface/scene_builder_test.dart:544]: we also need a test the zooms the picture in and makes sure that the growth in pixel count stops at screen size (i.e. that we're not allocating enormous canvases)

File: lib/web_ui/lib/src/engine/html/picture.dart [hunk 6]
```diff
@@ -603,3 +630,44 @@ class PersistedPicture extends PersistedLeafSurface {
...
+  final double xp = (m[0] * x) + (m[4] * y) + m[12];
+  final double yp = (m[1] * x) + (m[5] * y) + m[13];
+  final double wp = 1.0 / ((m[3] * x) + (m[7] * y) + m[15]);
+  double scaleX = ((xp * wp) - topLeftX).abs() / width;
+  double scaleY = ((yp * wp) - topLeftY).abs() / height;
+  double scale = math.min(scaleX, scaleY);
+  if (scale < kEpsilon || scale == 1) {
+    // Handle local paint bounds scaled to 0, typical when using
+    // transform animations.
+    return 1.0;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/picture.dart:655]: Shouldn't this be `if ((scale - 1).abs() < kEpsilon)`? Also scale 0 probably means "empty picture".

File: lib/web_ui/lib/src/engine/html/picture.dart [hunk 7]
```diff
@@ -603,3 +630,44 @@ class PersistedPicture extends PersistedLeafSurface {
...
+  }
+  if (scale > 1) {
+    // Normalize scale to multiples of 2: 1x, 2x, 4x, 6x, 8x.
+    // This is to prevent frequent rescaling of canvas during animations.
+    //
+    // On a fullscreen high dpi device dpi*density*resolution will demand
+    // too much memory, so clamp at 4.
+    scale = math.min(4.0, ((scale / 2.0).ceil() * 2.0));
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/picture.dart:663]: Should we normalize before the previous `if` statement? This way if `scale` snaps to 1x we just return 1.

File: lib/web_ui/lib/src/engine/html/surface.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/surface.dart:828]: Yes. Removed.

File: lib/web_ui/test/engine/surface/scene_builder_test.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/test/engine/surface/scene_builder_test.dart:30]: Done.

File: lib/web_ui/test/engine/surface/scene_builder_test.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/test/engine/surface/scene_builder_test.dart:532]: Done.

File: lib/web_ui/test/engine/surface/scene_builder_test.dart (hunk 4 above)
Comment (ferhatb) [reply lib/web_ui/test/engine/surface/scene_builder_test.dart:529]: Updated comment. thx.

File: lib/web_ui/test/engine/surface/scene_builder_test.dart (hunk 5 above)
Comment (ferhatb) [reply lib/web_ui/test/engine/surface/scene_builder_test.dart:544]: When you zoom out, the big canvas becomes small ie scale factor < 1. I added another test case for zoom in.

File: lib/web_ui/lib/src/engine/html/picture.dart (hunk 6 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/picture.dart:655]: Added and updated comment. It guards the code below from divide by zero.

File: lib/web_ui/lib/src/engine/html/picture.dart (hunk 7 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/picture.dart:663]: I like the early exit. For many apps most of the time scale is just 1.0 unless you use boxfit.

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**